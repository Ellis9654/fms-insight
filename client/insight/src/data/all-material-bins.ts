/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import * as im from "immutable";

export type MaterialList = ReadonlyArray<Readonly<api.IInProcessMaterial>>;

export type AllMaterialBins = im.Map<string, MaterialList>;

export function selectAllMaterialIntoBins(curSt: Readonly<api.ICurrentStatus>): AllMaterialBins {
  const palLoc = im
    .Map(curSt.pallets)
    .map(st => " (" + st.currentPalletLocation.group + " #" + st.currentPalletLocation.num.toString() + ")");

  return im
    .Seq(curSt.material)
    .map(mat => {
      switch (mat.location.type) {
        case api.LocType.InQueue:
          return { region: mat.location.currentQueue || "Queue", mat };
        case api.LocType.OnPallet:
          let region = "Pallet";
          if (mat.location.pallet) {
            region = "Pallet " + mat.location.pallet.toString() + palLoc.get(mat.location.pallet) || "";
          }
          return { region, mat };
        case api.LocType.Free:
        default:
          switch (mat.action.type) {
            case api.ActionType.Loading:
              return { region: "Raw Material", mat };
            default:
              return { region: "Free Material", mat };
          }
      }
    })
    .groupBy(x => x.region)
    .map(group =>
      group
        .map(x => x.mat)
        .sortBy(mat => {
          switch (mat.location.type) {
            case api.LocType.OnPallet:
              return mat.location.face;
            case api.LocType.InQueue:
            case api.LocType.Free:
            default:
              return mat.location.queuePosition;
          }
        })
        .valueSeq()
        .toArray()
    )
    .toMap();
}
