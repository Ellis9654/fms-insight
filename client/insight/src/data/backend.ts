/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import * as im from "immutable";
import { User } from "oidc-client";

export interface JobAPI {
  history(startUTC: Date, endUTC: Date): Promise<Readonly<api.IHistoricData>>;
  currentStatus(): Promise<Readonly<api.ICurrentStatus>>;
  mostRecentUnfilledWorkordersForPart(part: string): Promise<ReadonlyArray<Readonly<api.IPartWorkorder>>>;

  removeMaterialFromAllQueues(materialId: number): Promise<void>;
  setMaterialInQueue(materialId: number, queue: api.QueuePosition): Promise<void>;
  addUnprocessedMaterialToQueue(
    jobUnique: string,
    lastCompletedProcess: number,
    queue: string,
    pos: number,
    serial: string
  ): Promise<void>;
}

export interface ServerAPI {
  fMSInformation(): Promise<Readonly<api.IFMSInfo>>;
}

export interface LogAPI {
  get(startUTC: Date, endUTC: Date): Promise<ReadonlyArray<Readonly<api.ILogEntry>>>;
  recent(lastSeenCounter: number): Promise<ReadonlyArray<Readonly<api.ILogEntry>>>;
  logForMaterial(materialID: number): Promise<ReadonlyArray<Readonly<api.ILogEntry>>>;
  logForSerial(serial: string): Promise<ReadonlyArray<Readonly<api.ILogEntry>>>;
  getWorkorders(ids: string[]): Promise<ReadonlyArray<Readonly<api.IWorkorderSummary>>>;

  setInspectionDecision(inspType: string, mat: api.LogMaterial, inspect: boolean): Promise<Readonly<api.ILogEntry>>;
  recordInspectionCompleted(insp: api.NewInspectionCompleted): Promise<Readonly<api.ILogEntry>>;
  recordWashCompleted(insp: api.NewWash): Promise<Readonly<api.ILogEntry>>;
  setWorkorder(workorder: string, mat: api.LogMaterial): Promise<Readonly<api.ILogEntry>>;
  setSerial(serial: string, mat: api.LogMaterial): Promise<Readonly<api.ILogEntry>>;
}

export const BackendHost = process.env.NODE_ENV === "production" ? undefined : "localhost:5000";
const BackendUrl = BackendHost ? "http://" + BackendHost : undefined;
// tslint:disable-next-line:no-any
export const DemoMode = (window as any).FMS_INSIGHT_DEMO_MODE || false;

export let ServerBackend: ServerAPI = new api.ServerClient(BackendUrl);
export let JobsBackend: JobAPI = new api.JobsClient(BackendUrl);
export let LogBackend: LogAPI = new api.LogClient(BackendUrl);
let otherLogServers: ReadonlyArray<string> = [];
export let OtherLogBackends: ReadonlyArray<LogAPI> = [];

export function setOtherLogBackends(servers: ReadonlyArray<string>) {
  otherLogServers = servers;
  OtherLogBackends = servers.map(s => new api.LogClient(s));
}

export function setUserToken(u: User) {
  const token = u.access_token || u.id_token;
  function fetch(url: RequestInfo, init?: RequestInit) {
    return window.fetch(
      url,
      init
        ? { ...init, headers: { ...init.headers, Authorization: "Bearer " + token } }
        : { headers: { Authorization: "Bearer " + token } }
    );
  }
  ServerBackend = new api.ServerClient(BackendUrl, { fetch });
  JobsBackend = new api.JobsClient(BackendUrl, { fetch });
  LogBackend = new api.LogClient(BackendUrl, { fetch });
  OtherLogBackends = otherLogServers.map(s => new api.LogClient(s, { fetch }));
}

export interface MockData {
  readonly curSt: Readonly<api.ICurrentStatus>;
  readonly jobs: Readonly<api.IHistoricData>;
  readonly workorders: Map<string, ReadonlyArray<Readonly<api.IPartWorkorder>>>;
  readonly events: Promise<Readonly<api.ILogEntry>[]>;
}

function initMockBackend(data: Promise<MockData>) {
  ServerBackend = {
    fMSInformation() {
      return Promise.resolve({
        name: "mock",
        version: "1.0.0",
        requireScanAtWash: false,
        requireWorkorderBeforeAllowWashComplete: false,
        additionalLogServers: []
      });
    }
  };

  JobsBackend = {
    history(startUTC: Date, endUTC: Date): Promise<Readonly<api.IHistoricData>> {
      return data.then(d => d.jobs);
    },
    currentStatus(): Promise<Readonly<api.ICurrentStatus>> {
      return data.then(d => d.curSt);
    },
    mostRecentUnfilledWorkordersForPart(part: string): Promise<ReadonlyArray<Readonly<api.IPartWorkorder>>> {
      return data.then(d => d.workorders.get(part) || []);
    },

    removeMaterialFromAllQueues(materialId: number): Promise<void> {
      // do nothing
      return Promise.resolve();
    },
    setMaterialInQueue(materialId: number, queue: api.QueuePosition): Promise<void> {
      // do nothing
      return Promise.resolve();
    },
    addUnprocessedMaterialToQueue(
      jobUnique: string,
      lastCompletedProcess: number,
      queue: string,
      pos: number,
      serial: string
    ): Promise<void> {
      // do nothing
      return Promise.resolve();
    }
  };

  const serialsToMatId = data.then(d =>
    d.events.then(evts =>
      im.Map(
        im
          .Seq(evts)
          .filter(e => e.type === api.LogType.PartMark)
          .flatMap(e => e.material.map(m => [e.result, m.id] as [string, number]))
      )
    )
  );

  LogBackend = {
    get(startUTC: Date, endUTC: Date): Promise<ReadonlyArray<Readonly<api.ILogEntry>>> {
      return data.then(d =>
        d.events.then(evts =>
          im
            .Seq(evts)
            .filter(e => e.endUTC >= startUTC && e.endUTC <= endUTC)
            .toArray()
        )
      );
    },
    recent(lastSeenCounter: number): Promise<ReadonlyArray<Readonly<api.ILogEntry>>> {
      // no recent events, everything is static
      return Promise.resolve([]);
    },
    logForMaterial(materialID: number): Promise<ReadonlyArray<Readonly<api.ILogEntry>>> {
      return data.then(d =>
        d.events.then(evts =>
          im
            .Seq(evts)
            .filter(e => im.Seq(e.material).some(m => m.id === materialID))
            .toArray()
        )
      );
    },
    logForSerial(serial: string): Promise<ReadonlyArray<Readonly<api.ILogEntry>>> {
      return serialsToMatId.then(s => {
        var mId = s.get(serial);
        if (mId) {
          return this.logForMaterial(mId);
        } else {
          return Promise.resolve([]);
        }
      });
    },
    getWorkorders(ids: string[]): Promise<ReadonlyArray<Readonly<api.IWorkorderSummary>>> {
      // no workorder summaries
      return Promise.resolve([]);
    },

    setInspectionDecision(inspType: string, mat: api.LogMaterial, inspect: boolean): Promise<Readonly<api.ILogEntry>> {
      const evt = {
        counter: 0,
        material: [mat],
        pal: "",
        type: api.LogType.Inspection,
        startofcycle: false,
        endUTC: new Date(),
        loc: "Inspection",
        locnum: 1,
        result: inspect.toString(),
        program: "",
        elapsed: "00:00:00",
        active: "00:00:00",
        details: {
          InspectionType: inspType
        }
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    },
    recordInspectionCompleted(insp: api.NewInspectionCompleted): Promise<Readonly<api.ILogEntry>> {
      const evt: api.ILogEntry = {
        counter: 0,
        material: [insp.material],
        pal: "",
        type: api.LogType.InspectionResult,
        startofcycle: false,
        endUTC: new Date(),
        loc: "InspectionComplete",
        locnum: insp.inspectionLocationNum,
        result: insp.success.toString(),
        program: insp.inspectionType,
        elapsed: insp.elapsed,
        active: insp.active,
        details: insp.extraData
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    },
    recordWashCompleted(wash: api.NewWash): Promise<Readonly<api.ILogEntry>> {
      const evt: api.ILogEntry = {
        counter: 0,
        material: [wash.material],
        pal: "",
        type: api.LogType.Wash,
        startofcycle: false,
        endUTC: new Date(),
        loc: "Wash",
        locnum: wash.washLocationNum,
        result: "",
        program: "",
        elapsed: wash.elapsed,
        active: wash.active,
        details: wash.extraData
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    },
    setWorkorder(workorder: string, mat: api.LogMaterial): Promise<Readonly<api.ILogEntry>> {
      const evt: api.ILogEntry = {
        counter: 0,
        material: [mat],
        pal: "",
        type: api.LogType.OrderAssignment,
        startofcycle: false,
        endUTC: new Date(),
        loc: "OrderAssignment",
        locnum: 1,
        result: workorder,
        program: "",
        elapsed: "00:00:00",
        active: "00:00:00"
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    },
    setSerial(serial: string, mat: api.LogMaterial): Promise<Readonly<api.ILogEntry>> {
      const evt: api.ILogEntry = {
        counter: 0,
        material: [mat],
        pal: "",
        type: api.LogType.PartMark,
        startofcycle: false,
        endUTC: new Date(),
        loc: "Mark",
        locnum: 1,
        result: serial,
        program: "",
        elapsed: "00:00:00",
        active: "00:00:00"
      };
      return data.then(d =>
        d.events.then(evts => {
          evts.push(evt);
          return evt;
        })
      );
    }
  };
}

export function registerMockBackend() {
  const mockDataPromise = new Promise<MockData>(function(resolve: (d: MockData) => void) {
    // tslint:disable-next-line:no-any
    (window as any).FMS_INSIGHT_RESOLVE_MOCK_DATA = resolve;
  });
  initMockBackend(mockDataPromise);
}
