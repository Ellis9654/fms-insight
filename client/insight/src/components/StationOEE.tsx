/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import Grid from 'material-ui/Grid';
import * as im from 'immutable';
import * as numerable from 'numeral';
import { createSelector } from 'reselect';

import { Store } from '../data/store';
import { StationInUse } from '../data/events';
import { VictoryPie, VictoryLabel } from 'victory';

export interface StationOEEProps {
  station: string;
  oee: number;
}

export function StationOEE(p: StationOEEProps) {
  return (
    <svg viewBox="0 0 400 400">
      <VictoryPie
        standalone={false}
        width={400}
        height={400}
        data={[{x: 1, y: p.oee}, {x: 2, y: 1 - p.oee}]}
        innerRadius={140}
        cornerRadius={25}
        labels={() => ''}
        style={{
          data: { fill: (d: {x: number, y: number}) =>
            d.x === 1 ? '#795548' : '#E0E0E0'
          }
        }}
      />
      <VictoryLabel
        textAnchor="middle"
        verticalAnchor="middle"
        x={200}
        y={170}
        text={p.station}
        style={{fontSize: 45}}
      />
      <VictoryLabel
        textAnchor="middle"
        verticalAnchor="middle"
        x={200}
        y={230}
        text={numerable(p.oee).format('0.0%')}
        style={{fontSize: 30}}
      />
    </svg>
  );
}

export interface StationHours {
    readonly station: string;
    readonly hours: number;
}

export interface Props {
  station_active_hours_past_week: im.Seq<number, StationHours>;
  system_active_hours_per_week: number;
}

export function StationOEEs(p: Props) {
  return (
    <Grid container justify="space-around">
      {
        p.station_active_hours_past_week.map((stat, idx) => (
          <Grid item xs={6} sm={4} md={2} lg={1} key={idx}>
            <StationOEE station={stat.station} oee={stat.hours / p.system_active_hours_per_week}/>
          </Grid>
        ))
      }
    </Grid>
  );
}

export function stationHoursInLastWeek(use: im.List<StationInUse>): im.Seq<number, StationHours> {
    const m = new Map<string, number>();
    use.forEach(s => {
        m.set(s.station, (m.get(s.station) || 0) + s.hours);
    });
    return im.Seq(m)
        .map((v, k) => ({
            station: v[0],
            hours: v[1]
        }))
        .sortBy(e => e.station);
}

const oeeSelector = createSelector(
  (s: Store) => s.Events.last_week_of_hours,
  stationHoursInLastWeek
);

export default connect(
  (s: Store) => ({
    station_active_hours_past_week: oeeSelector(s),
    system_active_hours_per_week: s.Events.system_active_hours_per_week,
  })
)(StationOEEs);