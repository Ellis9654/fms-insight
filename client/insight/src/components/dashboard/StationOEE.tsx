/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import Grid from 'material-ui/Grid';
import * as im from 'immutable';
import * as numerable from 'numeral';
import { createSelector } from 'reselect';

import { Store } from '../../data/store';
import { StationInUse } from '../../data/events';

export interface StationOEEProps {
  station: string;
  oee: number;
}

function polarToCartesian(centerX: number, centerY: number, radius: number, angleInRadians: number) {
  return {
    x: centerX + (radius * Math.cos(angleInRadians)),
    y: centerY + (radius * Math.sin(angleInRadians))
  };
}

function describeArc(cx: number, cy: number, radius: number, startAngle: number, endAngle: number) {

    const start = polarToCartesian(cx, cy, radius, endAngle);
    const end = polarToCartesian(cx, cy, radius, startAngle);

    const largeArcFlag = endAngle - startAngle <= Math.PI ? '0' : '1';

    var d = [
        'M', start.x, start.y,
        'A', radius, radius, Math.PI / 2, largeArcFlag, '0', end.x, end.y
    ].join(' ');

    return d;
}

export function StationOEE(p: StationOEEProps) {
  const arcPoint = -Math.PI / 2 + 2 * Math.PI * p.oee;
  return (
    <svg viewBox="0 0 400 400">
      <path
        d={describeArc(200, 200, 150, -Math.PI / 2, arcPoint)}
        fill="transparent"
        stroke="#795548"
        strokeWidth={10}
      />
      <path
        d={describeArc(200, 200, 150, arcPoint, 1.5 * Math.PI)}
        fill="transparent"
        stroke="#E0E0E0"
        strokeWidth={10}
      />
      <text x={200} y={190} textAnchor="middle" style={{fontSize: 45}}>
        {p.station}
      </text>
      <text x={200} y={250} textAnchor="middle" style={{fontSize: 30}}>
        {numerable(p.oee).format('0.0%')}
      </text>
    </svg>
  );
}

export interface StationHours {
    readonly station: string;
    readonly hours: number;
}

export interface Props {
  station_active_hours_past_week: im.Seq<number, StationHours>;
  system_active_hours_per_week: number;
}

export function StationOEEs(p: Props) {
  return (
    <Grid container justify="space-around">
      {
        p.station_active_hours_past_week.map((stat, idx) => (
          <Grid item xs={6} sm={4} md={2} lg={1} key={idx}>
            <StationOEE station={stat.station} oee={stat.hours / p.system_active_hours_per_week}/>
          </Grid>
        ))
      }
    </Grid>
  );
}

export function stationHoursInLastWeek(use: im.List<StationInUse>): im.Seq<number, StationHours> {
    const m = new Map<string, number>();
    use.forEach(s => {
        m.set(s.station, (m.get(s.station) || 0) + s.hours);
    });
    return im.Seq(m)
        .map((v, k) => ({
            station: v[0],
            hours: v[1]
        }))
        .sortBy(e => e.station);
}

const oeeSelector = createSelector(
  (s: Store) => s.Events.last30.oee.last_week_of_hours,
  stationHoursInLastWeek
);

export default connect(
  (s: Store) => ({
    station_active_hours_past_week: oeeSelector(s),
    system_active_hours_per_week: s.Events.last30.oee.system_active_hours_per_week,
  })
)(StationOEEs);