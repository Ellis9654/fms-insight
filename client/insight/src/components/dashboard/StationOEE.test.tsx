/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as React from 'react';
import { StationOEE, StationOEEs, stationHoursInLastWeek } from './StationOEE';
import * as im from 'immutable';
import { shallow } from 'enzyme';

import * as api from '../../data/api';

it('calculates station hours', () => {
  const date = new Date();
  const hours = [
    {date, station: 'zzz', hours: 2},
    {date, station: 'abc', hours: 4},
    {date, station: 'abc', hours: 10},
    {date, station: 'zzz', hours: 5},
    {date, station: 'zzz', hours: 4},
    {date, station: 'abc', hours: 11},
  ];

  expect(stationHoursInLastWeek(im.List(hours)).toArray()).toEqual(
    [
      ["zzz", 2 + 5 + 4],
      ["abc", 4 + 10 + 11],
    ]
  );
});

it('displays station hours', () => {
  const hours = im.Map({
    "abc #1": 40,
    "zzz #1": 3,
  });
  const pals = im.Map({
    "aaa #1": {
      pal: {
        pallet: {
          pallet: "5",
          fixtureOnPallet: "",
          onHold: false,
          numFaces: 2,
          currentPalletLocation: new api.PalletLocation({
            loc: api.PalletLocationEnum.LoadUnload,
            group: "abc",
            num: 1
          }),
        },
        material: []
      },
    }
  });
  const val = shallow(
    <StationOEEs
      system_active_hours_per_week={100}
      station_active_hours_past_week={hours}
      pallets={pals}
    />);
  expect(val).toMatchSnapshot('station oee table');
});

it('displays a single station oee', () => {
  const pal = {
    pallet: {
      pallet: "7",
      fixtureOnPallet: "",
      onHold: false,
      numFaces: 2,
      currentPalletLocation: new api.PalletLocation({
        loc: api.PalletLocationEnum.LoadUnload,
        group: "aaa",
        num: 1,
      }),
    },
    material: [
      {
        materialID: 10,
        jobUnique: "aaa",
        partName: "aaa",
        process: 2,
        path: 1,
        signaledInspections: ["a", "b"],
        location: new api.InProcessMaterialLocation({
          type: api.LocType.OnPallet,
          pallet: "7",
          face: 1,
          queuePosition: 0,
        }),
        action: new api.InProcessMaterialAction({
          type: api.ActionType.Loading,
          loadOntoFace: 1,
          processAfterLoad: 2,
          pathAfterLoad: 1,
        }),
      }
    ]
  };

  const val = shallow(<StationOEE station="aaa" oee={0.43} pallet={pal}/>).dive().dive();
  expect(val).toMatchSnapshot('station oee gauge');
});

it('displays a single station without oee', () => {
  const val = shallow(<StationOEE station="bbb" oee={0.01}/>).dive().dive();
  expect(val).toMatchSnapshot('station empty guage');
});
