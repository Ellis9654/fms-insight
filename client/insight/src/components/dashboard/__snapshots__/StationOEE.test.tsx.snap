// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays a single station oee: station oee gauge 1`] = `
<WithStyles(Tooltip)
  classes={
    Object {
      "tooltip": "Component-tooltip-1",
    }
  }
  title={
    <UNDEFINED>
      <div>
        <span>
          Pallet
          : 
        </span>
        <span>
          7
        </span>
      </div>
      <div>
        <span>
          Part
          : 
        </span>
        <span>
          aaa-2 (loading)
        </span>
      </div>
    </UNDEFINED>
  }
>
  <svg
    viewBox="0 0 400 400"
  >
    <React.Fragment>
      <path
        d="M 263.86689373476094 335.72405786990294 A 150 150 1.5707963267948966 0 0 200 50"
        fill="transparent"
        stroke="#795548"
        strokeWidth={10}
      />
      <path
        d="M 199.99999999999997 50 A 150 150 1.5707963267948966 1 0 263.86689373476094 335.72405786990294"
        fill="transparent"
        stroke="#E0E0E0"
        strokeWidth={10}
      />
    </React.Fragment>
    <text
      style={
        Object {
          "fontSize": 45,
        }
      }
      textAnchor="middle"
      x={200}
      y={190}
    >
      aaa
    </text>
    <text
      style={
        Object {
          "fontSize": 30,
        }
      }
      textAnchor="middle"
      x={200}
      y={250}
    >
      Pal 7
    </text>
  </svg>
</WithStyles(Tooltip)>
`;

exports[`displays a single station without oee: station empty guage 1`] = `
<WithStyles(Tooltip)
  classes={
    Object {
      "tooltip": "Component-tooltip-1",
    }
  }
  title={
    <UNDEFINED>
      <div>
        <span>
          Pallet
          : 
        </span>
        <span>
          None
        </span>
      </div>
    </UNDEFINED>
  }
>
  <svg
    viewBox="0 0 400 400"
  >
    <circle
      cx={200}
      cy={200}
      fill="transparent"
      r={150}
      stroke="#E0E0E0"
      strokeWidth={10}
    />
    <text
      style={
        Object {
          "fontSize": 45,
        }
      }
      textAnchor="middle"
      x={200}
      y={190}
    >
      bbb
    </text>
    <text
      style={
        Object {
          "fontSize": 30,
        }
      }
      textAnchor="middle"
      x={200}
      y={250}
    >
      Empty
    </text>
  </svg>
</WithStyles(Tooltip)>
`;

exports[`displays station hours: station oee table 1`] = `
<WithStyles(GridWrapper)
  container={true}
  justify="space-around"
>
  <WithStyles(GridWrapper)
    item={true}
    key="aaa #1"
    lg={3}
    md={4}
    sm={6}
    xs={12}
  >
    <WithStyles(Component)
      oee={0}
      pallet={
        Object {
          "material": Array [],
          "pallet": Object {
            "currentPalletLocation": Object {
              "group": "abc",
              "loc": "LoadUnload",
              "num": 1,
            },
            "fixtureOnPallet": "",
            "numFaces": 2,
            "onHold": false,
            "pallet": "5",
          },
        }
      }
      station="aaa #1"
    />
  </WithStyles(GridWrapper)>
  <WithStyles(GridWrapper)
    item={true}
    key="abc #1"
    lg={3}
    md={4}
    sm={6}
    xs={12}
  >
    <WithStyles(Component)
      oee={0.4}
      station="abc #1"
    />
  </WithStyles(GridWrapper)>
  <WithStyles(GridWrapper)
    item={true}
    key="zzz #1"
    lg={3}
    md={4}
    sm={6}
    xs={12}
  >
    <WithStyles(Component)
      oee={0.03}
      station="zzz #1"
    />
  </WithStyles(GridWrapper)>
</WithStyles(GridWrapper)>
`;
